ext/DifferentiationInterfaceTapirExt/mutating.jl

```diff
@@ -0,0 +1,13 @@
+function DI.value_and_pullback!!(f!, y, dx, ::AutoTapir, x, dy, extras::Nothing)
+    rrule = build_rrule(f!, y, x)
+    dy_righttype = convert(typeof(y), dy)
+    dx_righttype = zero_sametype!!(dx, x)
```

Comment:
I'm already giving it the exact same types as `x` and `y` for the tangents, what more can I do?

Reply:
The correct thing to do here is to use `zero_tangent(x)` to generate `dx`.

I don't have a mutating version (I'm not sure how that would work anyway come to think of it), but if you're passing in `dx` and it to be mutated to contain the cotangent, you'll have to ensure that the correct type is passed in in the first place -- putting a `tangent_type(typeof(x)) == typeof(dx)` would suffice as far as Tapir is concerned.

Reply:
Basically I'd like to
- use `dx` if it has the right type
- create a new `zero_tangent` otherwise

Since conversion is a no-op if the type is right, I feel like this is the right approach